vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java

```diff
@@ -0,0 +1,243 @@
+            sendMessage(serverWebSocket, null, ApolloWSMessageType.CONNECTION_ACK);
+            break;
+          case CONNECTION_TERMINATE:
+            serverWebSocket.close();
+            break;
+          case START:
+            start(serverWebSocket, subscriptions, message);
+            break;
+          case STOP:
+            stop(serverWebSocket, subscriptions, opId);
+            break;
+          default:
+            sendError(serverWebSocket, opId, new Exception("Invalid message type!"));
+            break;
+        }
+      } catch (Exception e) {
+        sendError(serverWebSocket, null, e);
+      }
+    });
+
+    serverWebSocket.endHandler(v -> {
+      subscriptions.values().forEach(Subscription::cancel);
+
+      if (endHandler != null) {
+        endHandler.handle(serverWebSocket);
+      }
+    });
+  }
+
+  @Override
+  public ApolloWSHandler endHandler(Handler<ServerWebSocket> endHandler) {
+    this.endHandler = endHandler;
+
+    return this;
+  }
+
+  private void start(
+    ServerWebSocket serverWebSocket,
+    Map<String, Subscription> subscriptions,
+    JsonObject message
+  ) {
+    String opId = message.getString("id");
+
+    // Unsubscribe if it's subscribed
+    if (subscriptions.containsKey(opId)) {
+      stop(serverWebSocket, subscriptions, opId);
+    }
+
+    GraphQLQuery payload = message.getJsonObject("payload").mapTo(GraphQLQuery.class);
+    ExecutionInput.Builder builder = ExecutionInput.newExecutionInput();
+    builder.query(payload.getQuery());
+    String operationName = payload.getOperationName();
+    if (operationName != null) {
+      builder.operationName(operationName);
+    }
+    Map<String, Object> variables = payload.getVariables();
+    if (variables != null) {
+      builder.variables(variables);
+    }
+
+    graphQL.executeAsync(builder).thenAccept(executionResult -> {
+      if (executionResult.getData() instanceof Publisher) {
+        subscribe(serverWebSocket, subscriptions, opId, executionResult);
+      } else {
+        sendBackExecutionResult(serverWebSocket, opId, executionResult);
+      }
+    });
+  }
+
+  private void subscribe(
+    ServerWebSocket serverWebSocket,
+    Map<String, Subscription> subscriptions,
+    String opId,
+    ExecutionResult executionResult
+  ) {
+    Publisher<ExecutionResult> publisher = executionResult.getData();
+
+    AtomicReference<Subscription> subscriptionRef = new AtomicReference<>();
+    publisher.subscribe(new Subscriber<ExecutionResult>() {
+      @Override
+      public void onSubscribe(Subscription s) {
+        subscriptionRef.set(s);
+        subscriptions.put(opId, s);
+
+        s.request(1);
+      }
+
+      @Override
+      public void onNext(ExecutionResult er) {
+        sendMessage(serverWebSocket, opId, ApolloWSMessageType.DATA, er);
+
+        subscriptionRef.get().request(1);
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        sendError(serverWebSocket, opId, t);
```

Comment:
Shouldn't the subscription be removed here?

Reply:
You are right. In Observable documentation, talking about onError: "It will not make further calls to onNext or onCompleted"